flow/FileTraceLogWriter.cpp

```diff
@@ -78,13 +78,20 @@ void FileTraceLogWriter::write(const std::string& str) {
 			lastError(0);
 			remaining -= ret;
 			ptr += ret;
+			unsuccessfulFlushCount = 0;
 		} else {
+			unsuccessfulFlushCount++;
+			fprintf(stderr, "Unexpected error [%d] when flushing trace log.\n", errno);
```

Comment:
I don't know that this signal will work in all of the cases that the trace logging could fail. For example, I think that if the trace thread dies you may never actually even call this function. A number of the incidents I've looked at involved the trace thread dying, so we'll need to be able to capture that.

Also, the `FileTraceLogWriter` is not actually required for trace logging, though I don't think we have any alternatives at the moment. I think it's reasonable that a writer could report problems specific to its operation, but we should be able to detect more general problems too.

Reply:
> I don't know that this signal will work in all of the cases that the trace logging could fail. For example, I think that if the trace thread dies you may never actually even call this function. A number of the incidents I've looked at involved the trace thread dying, so we'll need to be able to capture that.

OK

> Also, the `FileTraceLogWriter` is not actually required for trace logging, though I don't think we have any alternatives at the moment. I think it's reasonable that a writer could report problems specific to its operation, but we should be able to detect more general problems too.

What about instead of reporting unsuccessful flush count, the `ITraceLogWriter` would have a function that can be called to fetch all `issues` it encountered?

Reply:
> A number of the incidents I've looked at involved the trace thread dying

How did you find out the thread dying? Is there a guard or single handler or something that can notice us about this?

Reply:
Unfortunately, no, there's nothing reporting that the trace thread is dying currently. See https://github.com/apple/foundationdb/issues/2506.

It looks like `FileTraceLogWriter` has a signaling mechanism for errors (its `onError` function), but that hasn't been generalized at all. Adding a function for getting issues to the interface seems reasonable to me to handle this.

Reply:
@ajbeamon I added a loose liveness check for the writer thread in commit 4a69b6f1ce3abdb66b9246b9d673f27bd20042a1. It's also possible to adopt the `pthread_kill(writer_thread, 0)` check if we need the accurate answer.